gh-119786: Add InternalDocs/qsbr.md. (#135411)
python/cpython: InternalDocs/qsbr.md

# Quiescent-State Based Reclamation (QSBR)

## Introduction

When implementing lock-free data structures, a key challenge is determining
when it is safe to free memory that has been logically removed from a
structure. Freeing memory too early can lead to use-after-free bugs if another
thread is still accessing it. Freeing it too late results in excessive memory
consumption.

Safe memory reclamation (SMR) schemes address this by delaying the free
operation until all concurrent read accesses are guaranteed to have completed.

> Quiescent-State Based Reclamation (QSBR) is an SMR scheme used in Python's

Reviewer: ```suggestion
Quiescent-State Based Reclamation (QSBR) is a SMR scheme used in Python's
```

free-threaded build to manage the lifecycle of shared memory.

QSBR requires threads to periodically report that they are in a quiescent
state. A thread is in a quiescent state if it holds no references to shared
objects that might be reclaimed. Think of it as a checkpoint where a thread
signals, "I am not in the middle of any operation that relies on a shared
resource." In Python, the eval_breaker provides a natural and convenient place
for threads to report this state.

## Use in Free-Threaded Python

While CPython's memory management is dominated by reference counting and a
tracing garbage collector, these mechanisms are not suitable for all data
structures. For example, the backing array of a list object is not individually

> reference-counted but may have a shorter lifetime than the PyListObject that

Reviewer 2: ```suggestion
reference-counted but may have a shorter lifetime than the `PyListObject` that
```

contains it. We could delay reclamation until the next GC run, but we want
reclamation to be prompt and to run the GC less frequently in the free-threaded
build, as it requires pausing all threads.

Many operations in the free-threaded build are protected by locks. However, for
performance-critical code, we want to allow reads to happen concurrently with
updates. For instance, we want to avoid locking during most list read accesses.
If a list is resized while another thread is reading it, QSBR provides the
mechanism to determine when it is safe to free the list's old backing array.

Specific use cases for QSBR include:

> * Dictionary keys (PyDictKeysObject) and list arrays (ob_item): When a

Reviewer 3: ```suggestion
* Dictionary keys (PyDictKeysObject) and list arrays (_PyListArray): When a
```

> * Dictionary keys (PyDictKeysObject) and list arrays (`_PyListArray`): When a

Reviewer 2: ```suggestion
* Dictionary keys (`PyDictKeysObject`) and list arrays (`_PyListArray`): When a
```

dictionary or list that may be shared between threads is resized, we use QSBR
to delay freeing the old keys or array until it's safe. For dicts and lists
that are not shared, their storage can be freed immediately upon resize.

* Mimalloc mi_page_t: Non-locking dictionary and list accesses require
cooperation from the memory allocator. If an object is freed and its memory is
reused, we must ensure the new object's reference count field is at the same
memory location. In practice, this means when a mimalloc page (mi_page_t)
becomes empty, we don't immediately allow it to be reused for allocations of a
different size class. QSBR is used to determine when it's safe to repurpose the
page or return its memory to the OS.

## Implementation Details

### Core Implementation

> The proposal to add QSBR to Python is contained in Github issue 115103 [1].

Reviewer 3: ```suggestion
The proposal to add QSBR to Python is contained in Github issue 115103 [^1].
```
You can use github's footnotes for it

> The proposal to add QSBR to Python is contained in Github issue 115103 [^1].

Reviewer 2: ```suggestion
The proposal to add QSBR to Python is contained in [Github issue 115103](https://github.com/python/cpython/issues/115103).
```

Many details of that proposal have been copied here, so they can be kept
up-to-date with the actual implementation.

Python's QSBR implementation is based on FreeBSD's "Global Unbounded

> Sequences." [2, 3, 4].  It relies on a few key counters:

Reviewer 2: ```suggestion
Sequences." [^2, ^3, ^4].  It relies on a few key counters:
```

### Deferred Advance Optimization

To reduce memory contention from frequent updates to the global `wr_seq`, its
advancement is sometimes deferred. Instead of incrementing `wr_seq` on every
reclamation request, each thread tracks its number of deferrals locally. Once
the deferral count reaches a limit (QSBR_DEFERRED_LIMIT, currently 10), the
thread advances the global `wr_seq` and resets its local count.

When an object is added to the deferred-free list, its qsbr_goal is set to
`wr_seq` + 2. By setting the goal to the next sequence value, we ensure it's safe
to defer the global counter advancement. This optimization improves runtime
speed but may increase peak memory usage by slightly delaying when memory can
be reclaimed.

## Limitations

Determining the `rd_seq` requires scanning over all thread states. This operation
could become a bottleneck in applications with a very large number of threads
(e.g., >1,000). Future work may address this with more advanced mechanisms,
such as a tree-based structure or incremental scanning. For now, the
implementation prioritizes simplicity, with plans for refinement if
multi-threaded benchmarks reveal performance issues.

## References

> 1. https://github.com/python/cpython/issues/115103

Reviewer 3: ```suggestion
[^1]: https://github.com/python/cpython/issues/115103
```
Same for others

2. https://youtu.be/ZXUIFj4nRjk?t=694
3. https://people.kernel.org/joelfernandes/gus-vs-rcu

> 4. http://bxr.su/FreeBSD/sys/kern/subr_smr.c#44

Reviewer 3: ```suggestion
[^2]:. https://youtu.be/ZXUIFj4nRjk?t=694
[^3]: https://people.kernel.org/joelfernandes/gus-vs-rcu
[^4]: http://bxr.su/FreeBSD/sys/kern/subr_smr.c#44
```
